Merge upstream v1.12.0

tomtseng opened the pull request:
(This PR should not be squashed.)

Biggest merge conflict is that `train.py` has been rewritten to use PyTorch instead of TF, so most of our `train.py` changes got discarded except for the `disable-vtimeloss` flag.
The switch to PyTorch causes another issue: we can't warmstart from models trained with KataGo v1.11 or earlier since the training weight format has changed.

Changes we'll need to make to `go_attack` (implemented in https://github.com/AlignmentResearch/go_attack/pull/118):
* Add `torch` and `requests_toolbelt` to `compose/python/requirements.txt` so that the Python image used by our linting CircleCI job can `pytype` the `KataGo-custom` Python code without complaining about missing libraries
* At least for now, we'll still want to keep the TF training code around for warmstarting on old models — either keep a old branch of KataGo in `engines/` or copy the old TF training code into a separate directory `KataGo-custom` 
* If we want to finetune the new architectures we'll need to edit `kubernetes/train.sh` since some of the flags to `train.py` have changed (e.g., model kind is a flag `-model-kind` instead of a positional argument, and warmstarting is done with the `-initial-checkpoint` flag instead of done by copying files into a `initial_weights` directory)

lightvector committed:
Allow outputting npzs in shuffle

lightvector committed:
Cleanup old code

lightvector committed:
Run training from dated archive

lightvector committed:
Remove unneeded import

lightvector committed:
Pytorch training

lightvector committed:
Fixes, reduce self-stdev prediction loss coeff

lightvector committed:
Fixes for batch norm, use batch norm

lightvector committed:
More weight decay and shorter warmup

lightvector committed:
Various lr changes and such

lightvector committed:
Fixes for wd scaling

lightvector committed:
Lift bnorm params to config, push config loading later with more checks

lightvector committed:
Batch renorm

lightvector committed:
Prepare soft policy loss

lightvector committed:
Enable soft policy aux

lightvector committed:
Brenorm, fixes

lightvector committed:
Faster renorm stas, soft policy weight scale

lightvector committed:
Bottleneck architecture

lightvector committed:
Update test, other fixes

lightvector committed:
More fixes, rework bottleneck

lightvector committed:
Long long bottleneck

lightvector committed:
Attention pool stuff

lightvector committed:
Nested bottleneck, intermediate heads

lightvector committed:
Nested nested resblock

lightvector committed:
Big refactor and code cleanup for blocks

lightvector committed:
Attention pool experiments

lightvector committed:
Intermediate head testing, soft policy target work

lightvector committed:
Repvgg and gamma

lightvector committed:
Repvgg fixes and gamma fixes

lightvector committed:
More activation functions

lightvector committed:
Fix gpool activation

lightvector committed:
Various fixes and refactors

lightvector committed:
Fixscale

lightvector committed:
Fixbrenorm, fix fixscale init bug

lightvector committed:
Make bn lrs to match fixup, remove bn from heads, fix l2 reg for last batchnorm gamma, change adaptive l2 scale, fson

lightvector committed:
Batchnorm head

lightvector committed:
Fix unowned moving prop, map location

lightvector committed:
Val fixes and new configs

lightvector committed:
Adjust net size

lightvector committed:
Record in val metrics the train nsamp and weights

lightvector committed:
Some qol improvements for test/validation

lightvector committed:
Some comments, update contributors

lightvector committed:
Merge master to pytorch branch

lightvector committed:
Remove cuda v3in tensor desc, begin support for multiple activation funcs

lightvector committed:
Cudnn use manager to simplify desc management, remove dilated support

lightvector committed:
Clean up more cuda backend stuff

lightvector committed:
Cuda simple allocator some bufs

lightvector committed:
Simple allocator more stuff

lightvector committed:
Cuda move some stuff to separate utils file

lightvector committed:
Remove dilated res block support

lightvector committed:
Support nested bottleneck in desc

lightvector committed:
Cuda normactconv and use act from desc instead of assuming

lightvector committed:
Implemented nested bottleneck for cuda

lightvector committed:
Cuda finish prep for more activations

lightvector committed:
Cuda kernels simplify fp16 preprocessor

lightvector committed:
Cuda and desc Mish activations

lightvector committed:
Allow adjusting value loss coeffs

lightvector committed:
Export model pytorch

lightvector committed:
Preserve old behavior for td value loss scales default

lightvector committed:
Fix epsilon bn issue

lightvector committed:
Update contributors for source of neural net training ideas

lightvector committed:
Various fixes to allow comparison with cuda impl, make play work

lightvector committed:
Breaking change - change how max pooling is done to be padding-friendly with negative activations

lightvector committed:
Fix intermediate value debugging

lightvector committed:
Perform makedirs earlier so that the logfile has somewhere to go

lightvector committed:
Refactor search explore code and remove total child weight adjust for multithread

lightvector committed:
Fix up global pooling for mish

lightvector committed:
Some debug code

lightvector committed:
Revert breaking change in max pooling

This reverts commit 09e9a7ad5dce7f430e12e1da1fa08f36be45c6f1.

lightvector committed:
Add sleep to shuffle script just in case

lightvector committed:
Remove most tensorflow stuff

lightvector committed:
Reduce shuffle memory usage for targets

lightvector committed:
Better exclude logic

lightvector committed:
Adjust loop periods for distributed

lightvector committed:
FP16 and distributed training

lightvector committed:
Merge branch 'master' into gpu-backend-cleanup

lightvector committed:
Fix closing files at end of train

lightvector committed:
Fix bug with train bucket init and tracking, remove redundant save stuff, gnorm debug

lightvector committed:
Big cleanup of training scripts, ckpt is now the only stateful file for training

lightvector committed:
Remove unused code, add missing load_model

lightvector committed:
Edit checkpoint script

lightvector committed:
Various fixes

lightvector committed:
Adjust weight decay strength, fix gamma weight decay bug, optional guard against repeat data files

lightvector committed:
More robust recomputation of modelnorm_normal_baseline

lightvector committed:
Rearrange model loading so norms get computed if deleted from train state

lightvector committed:
Update allowed play versions, shuffle validation set as well

lightvector committed:
Double v1 channels for bigger nets

lightvector committed:
Some commented out debug lines for model

lightvector committed:
Update how swa works, minor fixes

lightvector committed:
Implement lookahead optimizer

lightvector committed:
Fix stat accum bug

lightvector committed:
Implement mish activations for opencl and eigen, fixed

lightvector committed:
Cuda xSize -> nnXLen

lightvector committed:
Eigen and OpenCL xSize -> nnXLen

lightvector committed:
Additional minor cuda cleanup

lightvector committed:
NormActConv and SimpleAllocator cleanups for opencl and eigen

lightvector committed:
Fix cuda gpooling bug introduced when adding new activations

lightvector committed:
Cuda const

lightvector committed:
Cuda and OpenCL and Eigen const

lightvector committed:
Eigen and OpenCL nested residual blocks

lightvector committed:
Restore old eigen floating point behavior

lightvector committed:
Merge branch 'master' into gpu-backend-cleanup

lightvector committed:
Add option to randomize val file order

lightvector committed:
Guards against ultrarare policy 0s due to hash collisions

lightvector committed:
Fix opencl include dirs cmake

lightvector committed:
Update model versions comment

lightvector committed:
More work on hintpos generation

lightvector committed:
Fix board placements bug

lightvector committed:
Add command to test gpu error on fp16 vs fp32

lightvector committed:
Attempt to implement new architecture for tensorrt

Commit:
Revise GTP configuration documentation

lightvector committed:
Fix tensorrt impl

lightvector committed:
Fix duplicate layer names in exported model file

lightvector committed:
Fix packaging version import

lightvector committed:
Display trt fatal errors to cerr

OmnipotentEntity committed:
Added missing gtp extension commands to documentation. (#701)

lightvector committed:
Move next docs back to the main docs since they're already implemented

lightvector committed:
Add missing cuda architecture to cmake

lightvector committed:
Also report max errors

lightvector committed:
Perform mish in fp32 for opencl

lightvector committed:
Add pruned root values test

lightvector committed:
Fix turn number computation for poses

lightvector committed:
Add 40b test

lightvector committed:
Fix weighting oddity in pruned root values

lightvector committed:
More board placements bugfixes, add tests

lightvector committed:
Fix hgemm wmma SB implementation

lightvector committed:
Add better checking for openCL dimensions of things

lightvector committed:
Work on hgemm implementation for 1x1 conv

lightvector committed:
Finish opencl hgemm 1x1

lightvector committed:
Vectored unchecked load for hgemm_wmma_nchw

lightvector committed:
Tighter error tolerance in opencl tuner

lightvector committed:
More rigorous opencl tuning

lightvector committed:
Fix issue saving hgemm wmma nchw

lightvector committed:
Simplify with direct asst instead of vloadvstore

lightvector committed:
Remove bad const qualifier, adjust error in opencl

lightvector committed:
Work on fp16 error gpu test and automate for contrib

lightvector committed:
Fix opencl enable fp16 tensor cores 1x1 bug

lightvector committed:
Fix null nneval bug

lightvector committed:
Merge new architecture support!

lightvector committed:
Merge pytorch training of new architecture

lightvector committed:
Fix various issues with training loop using new code

lightvector committed:
Fix length limit for hgemm_wmma_nchw

lightvector committed:
Update cuda version support logic in cmake

lightvector committed:
Update outdated version reference in docs

simon300000 committed:
Update .gitignore (#720)

lightvector committed:
Support bigger komi stdev

lightvector committed:
Support smoother komi distribution via interp to 0

lightvector committed:
Add support for pos metadata

lightvector committed:
Fix possible issue where the komi limits fed to the net may be too tight

lightvector committed:
Implement terminate_all for analysis engine (#727)

lightvector committed:
Add default true option to log errors and warnings in analysis engine (#728)

lightvector committed:
Merge

lightvector committed:
Update tests

lightvector committed:
Revise GTP some more

lightvector committed:
Update contributors

lightvector committed:
Version 1.12.0

lightvector committed:
Make gpu test lighter-weight for contribute

lightvector committed:
Fix some compile warnings

lightvector committed:
Update contributors

tomtseng committed:
Merge tag 'v1.12.0' into tomtseng/merge-v1.12.0

tomtseng committed:
v1.12.0 use_vtimeloss

tomtseng committed:
v1.12.0: pytype

tomtseng committed:
v1.12.0: pytype ignore torch.cuda

tomtseng committed:
v1.12.0: codespell

ed1d1a8d commented on cpp/tests/testamcts.cpp:
```diff
@@ -555,21 +555,23 @@ void AMCTSTests::checkFinalMoveSelection(const Search& bot) {
       double fpuValue = bot.getFpuValueForChildrenAssumeVisited(
           *tree.root, tree.root->nextPla, true, 1.0, parentUtility,
           parentWeightPerVisit, parentUtilityStdevFactor);
+      double exploreScaling = bot.getExploreScaling(totalChildWeight, parentUtilityStdevFactor);
```
```suggestion
      const double exploreScaling = bot.getExploreScaling(totalChildWeight, parentUtilityStdevFactor);
```

ed1d1a8d commented on cpp/tests/testamcts.cpp:
```diff
@@ -555,21 +555,23 @@ void AMCTSTests::checkFinalMoveSelection(const Search& bot) {
+          parentUtility, parentWeightPerVisit, false,
           false, maxChildWeight, NULL);
     }();
+    double exploreScaling = bot.getExploreScaling(totalChildWeight, 1.0);
```
```suggestion
    const double exploreScaling = bot.getExploreScaling(totalChildWeight, 1.0);
```

ed1d1a8d committed:
Merge branch 'christmas-merge-rebased' into tomtseng/merge-v1.12.0

ed1d1a8d committed:
Update handicap.txt to redact time taken

ed1d1a8d reviewed (changes requested):
Have checked over all non python/* files and those look good.

For the python/* files, since the pytorch code seems to be non backwards compatible with the tensorflow code (per lightvector's checkpoint comments), I'm thinking we just copy over the python/* files exactly as is from the v1.12.0 code. If we need to finetune an older network, we can just use an older version of the repo.

tomtseng committed:
Update cpp/tests/testamcts.cpp

Co-authored-by: ttw <1733389+ed1d1a8d@users.noreply.github.com>

tomtseng committed:
Update cpp/tests/testamcts.cpp

Co-authored-by: ttw <1733389+ed1d1a8d@users.noreply.github.com>

tomtseng commented:
(Context to make this conversation clearer to others: We now cannot warmstart from old <=v1.11 models since they're saved as TF checkpoints whereas v1.12.0 has switched to PyTorch and expects PyTorch checkpoints. I asked on Discord whether there was a nice way to convert from TF checkpoints to PyTorch checkpoints and lightvector said there's no official way. Someone else suggested trying [MMdnn](https://github.com/Microsoft/MMdnn) for conversion.)

The Python files were essentially just copied over except for `-disable-vtimeloss` plus some changes to make the Python linter happy. If we don't think we'll care about `-disable-vtimeloss` for finetuning the newer KataGo architectures then yeah, let's just copy the v1.12.0 Python code without changes and disable linting for those files. 
I think removing `-disable-vtimeloss` is fine — the newer KataGo architectures were all trained with vtimeloss and it seems consistent to finetune them with vtimeloss as well. But we used `-disable-vtimeloss` for training our adversaries, and we might care about that eventually (e.g., if we work on adding vision transformers and think it's easier to write them in PyTorch instead of TF).

It does seem like converting checkpoints from TF to PyTorch will take some effort and it's risky to devote time to it right now. 
I think I'll add another copy of KataGo in `/go_attack/engines` that has the old TF training code, since we'll be using it frequently. In particular, whenever we want to finetune our cyclic adversary, we're using its TF checkpoint. I'll spend today setting up a finetuning run of the cyclic adversary vs. a b18c384nbt net and see if anything breaks.

tomtseng committed:
searchresults: fix merge conflict

ed1d1a8d commented:
I'd actually vote the leave out vtimeloss completely. We can make a commit to add it back in once we start running experiments that actually require disabling vtimeloss. I'm afraid of adding it back in right now without proper testing.

tomtseng committed:
Revert adding --disable-vtimeloss

tomtseng commented:
OK, I removed the `-disable-vtimeloss` flag and made a draft PR in case we want to add the flag back in: https://github.com/AlignmentResearch/KataGo-custom/pull/98

In the corresponding `go_attack` PR https://github.com/AlignmentResearch/go_attack/pull/118 I made the old TF training code a separate submodule. I think I like this a little more because it preserves the Git history of the TF training files (easier to Git Blame)

tomtseng committed:
Merge branch 'stable' into tomtseng/merge-v1.12.0

tomtseng committed:
train.py: Add -disable-vtimeloss to PyTorch training (#98)

With the old TF training code we had a flag `-disable-vtimeloss` to disable the variance time loss. We removed it when we moved to KataGo v1.12.0, which changes the training code to PyTorch, since we didn't want to add extraneous changes to the PyTorch code and we weren't using the PyTorch code.

This PR adds the `-disable-vtimeloss` flag back.

Testing: Checked that running `train.py` with `-disable-vtimeloss` drives the loss down without driving vtimeloss down.

ed1d1a8d reviewed (approved):
Basically lgtm.

One clarification question -- I saw that `python/train.py` had number of changes compared to upstream. I'm guessing most of these were done to make linting pass? If so then this is good to merge IMO. If not then maybe we should revert the changes to make the file to match upstream more...

tomtseng committed:
train: Add TensorBoard logging (#103)

Changes:
* Add tensorboard logging back into the PyTorch training code

tomtseng committed:
Merge branch 'stable' into tomtseng/merge-v1.12.0

tomtseng committed:
Merge branch 'stable' into tomtseng/merge-v1.12.0
